### PR TITLE
feat: add `RegularArray._reduce_next` implementation (backport)

### PR DESCRIPTION
**Pull request by agoose77**
Backport of #1811 

- [ ] updated tests

### CODE REVIEW

File: src/awkward/_v2/contents/indexedoptionarray.py [hunk 1]
```diff
@@ -1397,58 +1394,14 @@ def _reduce_next(
         if not branch and negaxis == depth:
             return out
         else:
-            if isinstance(out, ak._v2.contents.RegularArray):
-                out = out.toListOffsetArray64(True)
-
             # If the result of `_reduce_next` is a list, and we're not applying at this
             # depth, then it will have offsets given by the boundaries in parents.
             # This means that we need to look at the _contents_ to which the `outindex`
             # belongs to add the option type
```
Comment (jpivarski) [inline src/awkward/_v2/contents/indexedoptionarray.py]: Same as https://github.com/scikit-hep/awkward/pull/1811/files#r1000924785

```suggestion
```

File: src/awkward/_v2/contents/regulararray.py [hunk 2]
```diff
@@ -1000,70 +1000,198 @@ def _reduce_next(
...
-        )
 
-        if not self._content.dimension_optiontype:
-            branch, depth = self.branch_depth
-            convert_shallow = negaxis == depth
-            convert_deep = negaxis + 2 == depth
+        branch, depth = self.branch_depth
+        nextlen = len(self) * self._size
+
+        if negaxis == depth and not branch:
```
Comment (jpivarski) [inline src/awkward/_v2/contents/regulararray.py]: Same as https://github.com/scikit-hep/awkward/pull/1811/files#r1000929199

```suggestion
        if not branch and negaxis == depth:
```

**Review by jpivarski — APPROVED** [review]
I'm always impressed by how thorough your backports are!

I think you had one action left on #1811: adding kernel test data.

Beyond that, once #1811 is done and the test values are copied over, this #1813 is done, too. It can be marked ready for review (you already got a review) and merged.

**Comment by jpivarski** [comment]
As far as I can see, this is up-to-date with #1811. Neither has kernel tests, but I'd say they're optional for the PR to be merged.

**Comment by jpivarski** [comment]
I'll manually merge this once the tests are done. There's no auto-merge here because I couldn't add any tests to the branch protection. (They didn't show up in the drop-down menu.)

**Comment by agoose77** [comment]
Sounds good! I'm just removing the last vestiges of `represents_regular`

**Comment by jpivarski** [comment]
I was getting confused about which one was failing—I had thought it was the `main` one and was very confused.

If need be, the `cuda_kernels` can be mercilessly cut out of the legacy `main-v1` branch, since it's only for "v2 in 1.x", and it's an incomplete implementation that will be finished in the 2.x series. That may be overkill, considering that the build problem was somehow solved in `main`, but I'm giving you the option to punt on this: if it's not easy to make it work, you can make it go away.

**Comment by agoose77** [comment]
The error in the build here was just because I'd modified the `ListOffsetArrayOf<T>` constructor, but missed a couple of places where the flag was passed in positionally. I'm able to build locally, and the CI *should* work now.

**Comment by jpivarski** [comment]
Oh, I thought the warning about the CUDA source code directory was the reason the tests failed (with warnings-as-errors).

Since this is working now, I'll merge after it finishes all of the tests.

**Pull request merged**